experimental/ir/fdp.go

```diff
@@ -410,34 +400,30 @@ type syntheticNames map[intern.ID]struct{}
 
 func (sn *syntheticNames) generate(candidate string, message Type) string {
 	table := message.Context().intern
-	var names = *sn
 	if *sn == nil {
-		names = make(map[intern.ID]struct{})
-		*sn = names
-
-		for field := range seq.Values(message.Fields()) {
-			names[field.InternedName()] = struct{}{}
-		}
-		for extn := range seq.Values(message.Extensions()) {
-			names[extn.InternedName()] = struct{}{}
-		}
-		for oneof := range seq.Values(message.Oneofs()) {
-			names[oneof.InternedName()] = struct{}{}
-		}
-		for ty := range seq.Values(message.Nested()) {
-			name := ty.Name()
-			name = name[strings.LastIndex(name, ".")+1:]
-			names[table.Intern(name)] = struct{}{}
-		}
+		*sn = mapsx.CollectSet(iterx.Chain(
+			seq.Map(message.Fields(), Field.InternedName),
+			seq.Map(message.Extensions(), Field.InternedName),
+			seq.Map(message.Oneofs(), Oneof.InternedName),
+			iterx.FlatMap(seq.Values(message.Nested()), func(ty Type) iter.Seq[intern.ID] {
+				if !ty.IsEnum() {
+					return iterx.Of(ty.InternedName())
+				}
+				return iterx.Chain(
```

Comment:
It took me a second to grok this. How about a comment, about enum value scoping, yadda yadda.